include/pybind11/detail/internals.h

```diff
@@ -491,15 +547,66 @@ struct local_internals {
     std::forward_list<ExceptionTranslator> registered_exception_translators;
 };
 
+inline std::unique_ptr<local_internals> *&get_local_internals_pp() {
+#ifdef PYBIND11_SUBINTERPRETER_SUPPORT
+    if (get_num_interpreters_seen() > 1) {
+        // Internals is one per interpreter. When multiple interpreters are alive in different
+        // threads we have to allow them to have different internals, so we need a thread_local.
+        static thread_local std::unique_ptr<local_internals> *t_internals_pp = nullptr;
+        static thread_local PyInterpreterState *istate_cached = nullptr;
+        // Whenever the interpreter changes on the current thread we need to invalidate the
+        // internals_pp so that it can be pulled from the interpreter's state dict.  That is slow,
+        // so we use the current PyThreadState to check if it is necessary.  The caller will see a
+        // null return and do the fetch from the state dict or create a new one (as needed).
+        auto *tstate = get_thread_state_unchecked();
+        if (!tstate) {
+            istate_cached = nullptr;
+            t_internals_pp = nullptr;
+        } else if (tstate->interp != istate_cached) {
+            istate_cached = tstate->interp;
+            t_internals_pp = nullptr;
+        }
+        return t_internals_pp;
+    }
```

Comment:
This is an (almost) exact copy of the code in lines 321-339. Could you please try consolidating, to avoid duplicating this high non-trivial code?

Optional suggestion:

As a first step, I'd move the two implementations next to each other and make everything work again.

As the second step, I'd try to consolidate the two copies. I'm not sure what'll work best. I'd probably start by templating on `InternalsType` (`internals`, `local_internals`) and see how it goes.

Reply:
Done